File: 0046-permutations/0046-permutations.java
class Solution {
    List<List<Integer>> res;
    List<Integer> path;
    int[] used;
    public List<List<Integer>> permute(int[] nums) {
        res = new ArrayList<>();
        path = new ArrayList<>();
        used = new int[nums.length];
        backtracking(nums);
        return res;
    }

    private void backtracking(int[] nums) {
      if(path.size() == nums.length) {
          res.add(new ArrayList<Integer>(path));
          return;
      }

      for (int i = 0; i < nums.length; i++) {
          if (used[i] == 0) {
              path.add(nums[i]);
              used[i] = 1;
              backtracking(nums);
              used[i] = 0;
              path.remove(path.size() - 1);
          } 
      }
    }
}

File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const res = [];
    const permutation = [];
    const used = new Array(nums.length).fill(false);
    
    function backtrack(start) {
        if (permutation.length === nums.length) {
            res.push([...permutation]);
            return;
        }
        
        for (let i = 0; i < nums.length; i++) {
            if (!used[i]) {
                used[i] = true;
                permutation.push(nums[i]);
                backtrack(i + 1);
                permutation.pop();
                used[i] = false;
            }
        }
    }
    
    backtrack(0);
    return res;
};